crates/xilem_masonry/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+            }
+        } else {
+            eprintln!("Got action {action:?} for unknown widget. Did you forget to use `with_action_widget`?");
+        }
+    }
+}
+
+impl<E: Widget + StoreInWidgetMut> RootWidgetMut<'_, E> {
+    pub fn get_element(&mut self) -> WidgetMut<'_, E> {
+        self.ctx.get_mut(&mut self.widget.pod)
+    }
+}
+
+impl<State, Logic, View> Xilem<State, Logic, View>
+where
+    Logic: FnMut(&mut State) -> View,
+    View: MasonryView<State>,
+{
+    pub fn new(mut state: State, mut logic: Logic) -> Self {
+        let first_view = logic(&mut state);
+        let mut view_cx = ViewCx {
+            id_path: vec![],
+            widget_map: HashMap::new(),
+        };
+        let root_widget = RootWidget {
+            pod: first_view.build(&mut view_cx),
+        };
+        Xilem {
+            driver: MasonryDriver {
+                current_view: first_view,
+                logic,
+                state,
+                view_cx,
+            },
+            root_widget,
+        }
+    }
+
+    // TODO: Make windows a specific view
+    pub fn run_windowed(self, window_title: String) -> Result<(), EventLoopError>
+    where
+        State: 'static,
+        Logic: 'static,
+        View: 'static,
+    {
+        let event_loop = EventLoop::new().unwrap();
+        let window_size = LogicalSize::new(600., 800.);
+        let window = WindowBuilder::new()
+            .with_title(window_title)
+            .with_resizable(true)
+            .with_min_inner_size(window_size)
+            .build(&event_loop)
+            .unwrap();
+        self.run_windowed_in(window, event_loop)
+    }
+
+    // TODO: Make windows into a custom view
+    pub fn run_windowed_in(
+        self,
+        window: Window,
+        event_loop: EventLoop<()>,
+    ) -> Result<(), EventLoopError>
+    where
+        State: 'static,
+        Logic: 'static,
+        View: 'static,
+    {
+        EventLoopRunner::new(self.root_widget, window, event_loop, self.driver).run()
+    }
+}
+pub trait MasonryView<State, Action = ()>: Send + 'static {
+    type Element: Widget + StoreInWidgetMut;
+    fn build(&self, cx: &mut ViewCx) -> WidgetPod<Self::Element>;
+    fn message(
+        &self,
+        id_path: &[ViewId],
+        message: Box<dyn Any>,
+        app_state: &mut State,
+    ) -> MessageResult<Action>;
+    fn rebuild(
+        &self,
+        _cx: &mut ViewCx,
+        prev: &Self,
+        // _id: &mut Id,
+        element: WidgetMut<Self::Element>,
+    ) -> ChangeFlags;
+}
+
+#[must_use]
+pub struct ChangeFlags {
+    changed: bool,
+}
+
+impl ChangeFlags {
+    const CHANGED: Self = ChangeFlags { changed: true };
+    const UNCHANGED: Self = ChangeFlags { changed: false };
+}
+
+pub struct ViewCx {
+    /// The map from a widgets id to its position in the View tree.
+    ///
+    /// This includes only the widgets which might send actions
+    /// This is currently never cleaned up
+    widget_map: HashMap<WidgetId, Vec<ViewId>>,
```

Comment:
I believe we expect Masonry to do internal tree structure tracking at some point, which I think would include widget deleted events?

I don't know a *clean* way to tidy this up, otherwise